File: a2/casan/l2_154.py
"""
This module contains the subclasses of l2 and l2net for 802.15.4 networking.
"""
import sys
import os
import fcntl
import termios
import time

from casan import l2


class l2addr_154:
    """
    This class represents a 802.15.4 network address.
    """
    def __init__ (self, s=None):
        """
        Initialize the object from a string representation such as 'ca:fe'
        """
        if s is None:
            self.addr = None
        else:
            sl = s.split (':')
            self.addr = bytes ([int (ss, 16) for ss in sl])

    def __eq__ (self, other):
        """
        Test for equality between 2 l2addr_154 objects.
        """
        return False if other is None else self.addr == other.addr

    def __ne__ (self, other):
        """
        Test for difference between 2 l2addr_154 objects.
        """
        return not self.__eq__ (other)

    def __repr__ (self):
        """
        Return a printable representation of a l2addr_154 object.
        """
        rep = ''
        for b in self.addr:
            rep += hex (b) [2:]
        return rep

    def __str__ (self):
        """
        Return a human readable string representation of a l2addr_154.
        """
        rep = []
        for b in self.addr:
            atom = hex (b) [2:]
            if len (atom) == 1:
                atom = '0' + atom        # Prefix single-digits with 0
            rep.append (atom)
        return ':'.join (rep)


class l2net_154:
    """
    This class represents a L2 802.15.4 network connection.
    """
    #
    # Constants (Python makes them class variables)
    #
    HEADER_SIZE = 9
    FCS = 2
    MTU = 127
    XBEE_MTU = 100 + HEADER_SIZE + FCS
    XBEE_START = 0x7e
    XBEE_TX_SHORT = 0x01
    XBEE_MIN_FRAME_SIZE = 5
    XBEE_MAX_FRAME_SIZE = 115
    RX_SHORT_OPT_BROADCAST = 0x02
    XBEE_READ_MAX = 1000
    # XBee frame types
    XBEE_FT_TX_STATUS = 0x89
    XBEE_FT_RX_LONG = 0x80
    XBEE_FT_RX_SHORT = 0x81

    broadcast = l2addr_154 ('ff:ff:ff:ff:ff:ff:ff:ff')

    def __str__ (self):
        """
        Return a string representation of the network object.
        Used for debugging purposes.
        """
        s = 'Network type : 802.15.4\n'
        s += 'Interface : ' + self.iface + '\n'
        s += 'Interface address : ' + str (self.a) + '\n'
        s += 'PANID : ' + str (self.pan) + '\n'
        s += 'MTU : ' + str (self.mtu) + '\n'

    def __init__ (self):
        """
        Construct a l2net_154 object with some default values.
        """
        self.max_latency = 5
        self.fd = -1
        self.framelist = []
        self.buffer = bytearray ()
        self.iface = None

    def _init_xbee (self, iface, addr, panid, channel):
        """
        Initialize access to the Digi XBee chip
        """
        if not (0 < self.mtu <= self.XBEE_MTU):
            self.mtu = self.XBEE_MTU
        if not (10 < self.channel < 26):
            return False
        try:
            # Open and initialize device
            if not iface.startswith ('/dev'):
                iface = '/dev/' + iface
            self.fd = os.open (iface, os.O_RDWR)
            attrs = termios.tcgetattr (self.fd)
            attrs [:4] = [termios.IGNBRK | termios.IGNPAR,
                          0,
                          termios.CS8 | termios.CREAD | termios.B9600,
                          0]
            for attr in attrs [6]:
                # POSIX_VDISABLE may not be implemented in python, at
                # least I couldn't find any information about it, maybe
                # check in python stdlib source code?
                #attr = termios.POSIX_VDISABLE
                pass
            termios.tcsetattr (self.fd, termios.TCSANOW, attrs)

            def send_AT_command (s):
                """
                Sends an AT command over the serial port and checks
                for the returned value.
                """
                os.write (self.fd, s)
                r = bytearray ()
                while True:
                    c = os.read (self.fd, self.XBEE_READ_MAX)
                    r += c
                    if c == b'\r':
                        if r [-3:] == b'OK\r':
                            return
                        elif r [-6:] == b'ERROR\r':
                            raise RuntimeError ('Error sending AT command'
                                                + ' to XBEE.\nCommand : '
                                                + s.decode ())
                        else:
                            r = bytearray ()

            # Initialize API mode
            time.sleep (0.2)
            send_AT_command (b'+++')
            time.sleep (0.2)
            send_AT_command (b'ATRE\r')

            # Short address
            sa = (b'ATMY' +
                  bytes (hex (self.a.addr [1]) [2:], 'utf-8') +
                  bytes (hex (self.a.addr [0]) [2:], 'utf-8') +
                  b'\r')
            send_AT_command (sa)

            # PANID
            panid = (b'ATID' +
                     bytes (hex (self.pan.addr [1]) [2:], 'utf-8') +
                     bytes (hex (self.pan.addr [0]) [2:], 'utf-8') +
                     b'\r')
            send_AT_command (panid)

            # Channel
            chan = b'ATCH' + bytes (hex (channel) [2:], 'utf-8') + b'\r'
            send_AT_command (chan)

            # 802.15.4 w/ ACKs
            send_AT_command (b'ATMM2\r')

            # Enter API mode
            send_AT_command (b'ATAP1\r')

            # Quit AT command mode
            send_AT_command (b'ATCN\r')

        except Exception as e:
            sys.stderr.write (str (e) + '\n')
            sys.stderr.write ('Error : cannot open interface ' + iface)
            return False

        # configure the device to non-blocking mode for asyncio
        if self.asyncio:
            fcntl.fcntl (self.fd, fcntl.F_SETFL, os.O_NDELAY)

        return True

    def init (self, iface, type_, mtu, addr, panid, channel, asyncio=False):
        """
        Initialize a l2net_154 object, opens and sets up the network interface
        :param iface: interface to start. Full path (eg: '/dev/ttyUSB0') or
                    just device name (eg: 'ttyUSB0').
        :type  iface: string
        :param type_: type of the interface. Use 'xbee' here.
        :type  type_: string
        :param mtu: MTU for interface iface. If not valid, will be set to a
                    default value. 0 means default value.
        :type  mtu: integer
        :param addr: physical address to use for interface iface.
        :type  addr: string
        :param panid: PANID
        :type  panid: string
        :param channel: channel
        :type  channel: integer
        :param asyncio: True if this object is used by asyncio
        :type  asyncio: boolean
        :return: True if ok, False if error.
        """

        self.a = l2addr_154 (addr)
        self.pan = l2addr_154 (panid)
        self.channel = channel
        self.mtu = mtu
        self.asyncio = asyncio

        if type_ == 'xbee':
            r = self._init_xbee (iface, addr, panid, channel)
        else:
            raise RuntimeError ('Unknown type ' + str (type_))
        return r

    def term (self):
        """
        Close the connection.
        """
        os.close (self.fd)
        self.fd = -1

    def handle (self):
        """
        Return file handle, needed for asyncio
        :return: file handle
        """
        return self.fd

    def encode_transmit (self, destAddr, data):
        """
        Encode data into a valid 802.15.4 frame
        """
        dlen = 5 + len (data)
        cmdlen = 4 + dlen
        b = bytearray ()
        b.append (self.XBEE_START)
        b.append ((dlen & 0xff00) >> 8)
        b.append (dlen & 0x00ff)
        b.append (self.XBEE_TX_SHORT)
        b.append (0x41)
        b.append (destAddr.addr [1])
        b.append (destAddr.addr [0])
        b.append (0)
        b += data
        b.append (self.compute_checksum (b))
        return b

    @staticmethod
    def compute_checksum (buf):
        """
        Compute the checksum of a frame.
        """
        def i16_from_2i8 (i1, i2):
            """
            Returns a 16bits integer made from 2 8bits integers.
            The first integer will be the most significant byte of the result
            """
            return (i1 << 8) | i2
        paylen = i16_from_2i8 (buf [1], buf [2]) + 4
        c = 0
        for i in range (3, paylen - 1):
            c = (c + buf [i]) & 0xff
        return 0xff - c

    def send (self, dest_slave, data):
        """
        Send a frame on the network.
        :return: True if success, False either.
        """
        if len (data) <= self.MTU:
            cmd = self.encode_transmit (dest_slave.addr, data)
            clen = len (cmd)
            try:
                while clen > 0:
                    r = os.write (self.fd, cmd [:clen])
                    if r == 0:
                        return False
                    else:
                        clen = clen - r
                        cmd = cmd [r:]
                return True
            except Exception:
                return False

    def bsend (self, data):
        """
        Broadcast a frame on the network.
        """
        return self.send (self.broadcast, data)

    def recv (self):
        """
        Receive a frame.
        :return: tuple (PACKET_TYPE, SOURCE_ADDRESS, DATA)
                 PACKET_TYPE = see l2.PkType
        """
        r = (l2.PkType.PK_NONE, None, None)
        while r [0] is l2.PkType.PK_NONE:
            r = self.extract_received_packet ()
            if r [0] is l2.PkType.PK_NONE:
                x = self.read_complete_frame ()
                if not x:
                    break
        return r

    def extract_received_packet (self):
        """
        Extract a packet from the frame list.
        :return: tuple (PACKET_TYPE, (SOURCE_ADDRESS, DATA, LENGTH))
        """
        r = l2.PkType.PK_NONE
        a, data, = None, None
        for frame in self.framelist:
            self.framelist.remove (frame)
            if frame.type == self.XBEE_FT_RX_SHORT:
                a = l2addr_154 ()
                a.addr = bytes ([frame.rx_short.saddr & 0x00ff,
                                 (frame.rx_short.saddr & 0xff00) >> 8])
                data = frame.rx_short.data
                if frame.rx_short.options & self.RX_SHORT_OPT_BROADCAST:
                    r = l2.PkType.PK_BCAST
                else:
                    r = l2.PkType.PK_ME
                    break
        return (r, a, data)

    def read_complete_frame (self):
        """
        Read a full frame from the input buffer and adds it to the frame list.
        This function blocks until a complete frame is received or
        returns False when no input is available (in asyncio case)
        """
        found = False
        while not found:
            try:
                buf = os.read (self.fd, self.XBEE_READ_MAX)
            except BlockingIOError:
                # Exception raised when data is not available and
                # read set in non-blocking mode (i.e. asyncio mode)
                break
            if not buf:
                break
            self.buffer = self.buffer + buf
            while self.is_frame_complete ():
                self.extract_frame_to_list ()
                found = True

        return found

    def is_frame_complete (self):
        """
        Checks if there is a complete frame available for extraction from
        the input buffer.
        """
        complete = False
        while not complete:
            if self.XBEE_START not in self.buffer:
                # No start byte found
                break
            # Align start byte to buffer start
            self.buffer = self.buffer [self.buffer.index (self.XBEE_START):]

            # Is buffer long enough to extract a frame?
            buflen = len (self.buffer)
            if buflen < self.XBEE_MIN_FRAME_SIZE:
                break

            framelen = (self.buffer [1] << 8) | self.buffer [2]
            if framelen > self.XBEE_MAX_FRAME_SIZE:
                self.buffer = self.buffer [1:]
                continue

            # Got checksum?
            pktlen = framelen + 4
            if pktlen > buflen:
                break

            cksum = self.compute_checksum (self.buffer)
            if cksum != self.buffer [pktlen - 1]:
                self.buffer = self.buffer [1:]
                continue

            # At this point, we have a full valid frame in the buffer
            complete = True

        return complete

    def extract_frame_to_list (self):
        """
        Extract a complete frame from the input buffer and adds it to the
        list of received frames
        """
        class Frame:
            pass
        framelen = (self.buffer [1] << 8) | self.buffer [2]
        pktlen = framelen + 4
        f = Frame ()
        f.type = self.buffer [3]
        if f.type == self.XBEE_FT_TX_STATUS:
            f.tx_status = Frame ()
            f.tx_status.frame_id = self.buffer [4]
            f.tx_status.status = self.buffer [5]
        elif f.type == self.XBEE_FT_RX_SHORT:
            f.rx_short = Frame ()
            f.rx_short.saddr = (self.buffer [4] << 8) | self.buffer [5]
            f.rx_short.len = framelen - 5
            f.rx_short.data = bytes (self.buffer [8: f.rx_short.len + 8])
            f.rx_short.rssi = self.buffer [6]
            f.rx_short.options = self.buffer [7]
        else:
            sys.stderr.write ('PKT API ' + str (f.type) + ' unrecognized\n')
            # Return to avoid appending unknown frame
            return
        self.framelist.append (f)
        self.buffer = (bytearray ()
                       if len (self.buffer) == pktlen
                       else self.buffer [pktlen:])

File: c-master/casan/l2-154.cc
/**
 * @file l2-154.cc
 * @brief L2addr_154 and l2net_154 class implementations
 */

#include <iostream>
#include <cstring>

// #define	_BSD_SOURCE			// for termios on Linux

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

#include <termios.h>
#include <sys/ioctl.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>

#include "global.h"

#include "l2.h"
#include "l2-154.h"
#include "byte.h"

#define	XBEE_START		0x7e		// XBee API start delimiter
#define	XBEE_TX_SHORT		0x01		// XBee API command id

/*
 * MTU is expressed as the maximum MAC frame size, i.e. L2154_MTU bytes
 * (127 bytes). However, XBee limits payload data received and sent to
 * 100 bytes. So, the real MTU supported by XBee is:
 *	XBEE_MTU = 100 + (2+1+2+2+2)+2
 * since the controller adds the following informations:
 * - header:
 *	2 bytes : FCF (with the intra-PAN bit set)
 *	1 byte  : Sequence number
 *	2 bytes : destination address
 *	2 bytes : destination PANID
 *	2 bytes : source address (without PANID since the intra-PAN bit is set)
 * - trailer:
 *	2 bytes : FCS
 * We assume the use of 16-bits addresses, intra-PAN communication and
 * no encryption.
 *
 * Note: these numbers must match with slave encoding of frames (see
 * numbers in corresponding slave source file).
 */

#define	L2154_SIZE_HEADER	(2+1+2+2+2)
#define	L2154_SIZE_FCS		2

#define	XBEE_MTU		(100 + L2154_SIZE_HEADER + L2154_SIZE_FCS)

/* XBee API frame sizes */
#define	XBEE_MIN_FRAME_SIZE	5		// start:1+len:2+api:1+cksum:1
#define	XBEE_MAX_FRAME_SIZE	115		// RX 64bit addr, 100 bytes

namespace casan {

/******************************************************************************
 * l2addr_154 methods
 */

// default constructor
l2addr_154::l2addr_154 ()
{
}

/**
 * @brief Constructor with an address given as a string
 *
 * This constructor is used to initialize an address with a
 * string such as "`ca:fe`". 64-bits addresses are not supported.
 *
 * @param a address to be parsed
 */

l2addr_154::l2addr_154 (const char *a)
{
    int i = 0 ;
    byte b = 0 ;

    while (*a != '\0' && i < L2154ADDRLEN)
    {
	if (*a == ':')
	{
	    addr_ [i++] = b ;
	    b = 0 ;
	}
	else if (isxdigit (*a))
	{
	    byte x ;
	    char c ;

	    c = tolower (*a) ;
	    x = isdigit (c) ? (c - '0') : (c - 'a' + 10) ;
	    b = (b << 4) + x ;
	}
	else
	{
	    for (i = 0 ; i < L2154ADDRLEN ; i++)
		addr_ [i] = 0 ;
	    break ;
	}
	a++ ;
    }
    if (i < L2154ADDRLEN)
	addr_ [i++] = b ;

    /* Short addresses */
    if (i < L2154ADDRLEN)
    {
	int j ;

	/* report the i bytes of address to the end of the byte string */
	for (j = 0 ; j < i ; j++)
	    addr_ [L2154ADDRLEN - 1 - j] = addr_ [i - 1 - j] ;
	/* reset the L2154ADDRLEN - i first bytes to 0 */
	for (j = 0 ; j < L2154ADDRLEN - i ; j++)
	    addr_ [j] = 0 ;
    }
}

// copy constructor
l2addr_154::l2addr_154 (const l2addr_154 &l)
{
    *this = l ;
}

// copy assignment constructor
l2addr_154 &l2addr_154::operator= (const l2addr_154 &l)
{
    if (this != &l)
	*this = l ;
    return *this ;
}

/**
 * @brief 802.15.4 broadcast address
 */

l2addr_154 l2addr_154_broadcast ("ff:ff:ff:ff:ff:ff:ff:ff") ;

/**
 * @brief Print address
 *
 * This function is a hack needed for l2::operator<< overloading
 */

void l2addr_154::print (std::ostream &os) const
{
    for (int i = 0  ; i < L2154ADDRLEN ; i++)
    {
	if (i > 0)
	    os << ":" ;
	PRINT_HEX_DIGIT (os, addr_ [i] >> 4) ;
	PRINT_HEX_DIGIT (os, addr_ [i]     ) ;
    }
}

bool l2addr_154::operator== (const l2addr &other)
{
    l2addr_154 *oe = (l2addr_154 *) &other ;
    return std::memcmp (this->addr_, oe->addr_, L2154ADDRLEN) == 0 ;
}

bool l2addr_154::operator!= (const l2addr &other)
{
    return ! (*this == other) ;
}

/******************************************************************************
 * l2net_154 methods
 */

/**
 * @brief Initialize an IEEE 802.15.4 network access
 *
 * Initialize the IEEE 802.15.4 network access with needed constants.
 *
 * @param iface entry in /dev (with or without /dev) for XBee USB stick
 * @param type `xbee`
 * @param mtu user configured MTU or 0 for default network MTU
 * @param myaddr Our IEEE 802.15.4 address
 * @param panid PAN-id
 * @param channel channel number
 * @return -1 if initialization fails (`errno` is set)
 */

int l2net_154::init (const std::string iface, const char *type, const int mtu, const std::string myaddr, const std::string panid, const int channel)
{
    std::string dev ;
    int n = -1 ;			// default: init fails
    l2addr_154 a (myaddr.c_str ()) ;
    l2addr_154 pan (panid.c_str ()) ;	// same format as addr

    /* Various initializations */
    maxlatency_ = L2154MAXLATENCY ;

    /* Prepend /dev if needed */
    if (iface [0] == '/')
	dev = iface ;
    else
	dev = "/dev/" + iface ;

    if (strcmp (type, "xbee") == 0)
    {
	struct termios tm ;

	mtu_ = (mtu > 0 && mtu <= XBEE_MTU) ? mtu : XBEE_MTU ;
	if (channel < 11 || channel > 26)
	    return -1 ;

	/* Open device */
	fd_ = open (dev.c_str (), O_RDWR) ;
	if (fd_ == -1)
	    return -1 ;

	/* Initialize RS-232 driver */
	if (tcgetattr (fd_, &tm) == -1)
	{
	    close (fd_) ;
	    return -1 ;
	}
	tm.c_iflag = IGNBRK | IGNPAR ;
	tm.c_oflag = 0 ;
	tm.c_cflag = CS8 | CREAD | B9600 ;
	tm.c_lflag = 0 ;
	for (int i = 0 ; i < NTAB (tm.c_cc) ; i++)
	    tm.c_cc [i] = _POSIX_VDISABLE ;
	tm.c_cc [VMIN] = 1 ;		// read() waits for 1 char (at least)
	if (tcsetattr (fd_, TCSANOW, &tm) == -1)
	{
	    close (fd_) ;
	    return -1 ;
	}

	/* Initialize device to API (non-escaped) mode */
	sleep (2) ;
	write (fd_, "+++", 3) ;
	sleep (2) ;
	write (fd_, "ATRE\r", 5) ;		// restore factory defaults
	// write (fd_, "ATFR\r", 5) ;		// software reset
	// initialize short addresses, my address, panid
	{
	    char buf [MAXBUF] ;

	    // my short address : lowest significant byte first
	    std::sprintf (buf, "ATMY%02x%02x\r",
			a.addr_ [L2154ADDRLEN-1],
			a.addr_ [L2154ADDRLEN-2]) ;
	    write (fd_, buf, strlen (buf)) ;

	    // pan id : lowest significant byte first
	    std::sprintf (buf, "ATID%02x%02x\r",
			pan.addr_ [L2154ADDRLEN-1],
			pan.addr_ [L2154ADDRLEN-2]) ;
	    write (fd_, buf, strlen (buf)) ;

	    // channel
	    std::sprintf (buf, "ATCH%02x\r", channel) ;
	    write (fd_, buf, strlen (buf)) ;

	    // 802.15.4 with ACKs
	    write (fd_, "ATMM2\r", 6) ;
	}
	write (fd_, "ATAP1\r", 6) ;		// enter API mode
	write (fd_, "ATCN\r", 5) ;		// quit AT command mode
	
	n = 0 ;

	pbuffer_ = buffer_ ;
    }
    else
    {
	mtu_ = (mtu > 0 && mtu <= L2154MTU) ? mtu : L2154MTU ;
    }

    return n ;
}

/**
 * @brief Closes access to network
 */

void l2net_154::term (void)
{
    close (fd_) ;
}

/**
 * @brief Send a frame to the given destination address
 *
 * @return number of bytes sent
 */

int l2net_154::send (l2addr *daddr, void *data, int len)
{
    l2addr_154 *da = (l2addr_154 *) daddr ;
    int n ;
    byte cmd [MAXBUF] ;

    n = -1 ;
    if (len <= L2154MTU)
    {
	int cmdlen ;

	cmdlen = sizeof cmd ;
	if (encode_transmit (cmd, cmdlen, da, (byte *) data, len))
	{
	    byte *pcmd ;

	    /*
	     * Send encoded command, may be in multiple chunks
	     * since write() may return a value < cmdlen
	     */

	    n = 0 ;
	    pcmd = cmd ;
	    while (cmdlen > 0)
	    {
		int r ;

		r = write (fd_, pcmd, cmdlen) ;
		if (r == -1)
		{
		    n = -1 ;
		    break ;
		}
		else
		{
		    cmdlen -= r ;
		    pcmd += r ;
		    n += r ;
		}
	    }
	}
    }
    return n ;
}

/**
 * @brief Send a frame to the broadcast address
 *
 * @return number of bytes sent
 */

int l2net_154::bsend (void *data, int len)
{
    return send (&l2addr_154_broadcast, data, len) ;
}

// private method
bool l2net_154::encode_transmit (byte *cmd, int &cmdlen, l2addr_154 *daddr, byte *data, int len)
{
    byte *b ;
    int fdlen ;

    b = cmd ;
    // len of frame data (TX request) = tx:1+id:1+daddr:2+opt:1+data:len
    fdlen = 5 + len ;
    // len of encoded packet = start:1+len:2+frame:plen+cksum:1
    cmdlen = 4 + fdlen ;

    *b++ = XBEE_START ;
    *b++ = BYTE_HIGH (fdlen) ;
    *b++ = BYTE_LOW (fdlen) ;

    *b++ = XBEE_TX_SHORT ;
    // *b++ = 0 ;				// frame id
    *b++ = 0x41 ;				// frame id
    *b++ = daddr->addr_ [L2154ADDRLEN-1] ;	// lowest significant byte first
    *b++ = daddr->addr_ [L2154ADDRLEN-2] ;
    *b++ = 0 ;				// options
    std::memcpy (b, data, len) ;
    b += len ;

    *b++ = compute_checksum (cmd) ;

#if 0
    std::cout << "PKT=" ;
    for (int i = 0 ; i < cmdlen ; i++)
    {
	if (i > 0)
	    std::cout << ":" ;
	std::cout << std::hex << (int) cmd [i] ;
    }
    std::cout << "\n" << std::dec ;
#endif

    return true ;
}

// buf = encoded frame
int l2net_154::compute_checksum (const byte *buf)
{
    int c = 0 ;
    int paylen ;

    // length of whole API packet, from API start byte to checksum included
    paylen = INT16 (buf [1], buf [2]) + 4 ;

    for (int i = 3 ; i < paylen - 1 ; i++)
	c = (c + buf [i]) & 0xff ;
    return 0xff - c ;
}

/**
 * @brief Return the broadcast address for this network
 */

l2addr *l2net_154::bcastaddr (void)
{
    return &l2addr_154_broadcast ;
}

/**
 * @brief Receive frame
 *
 * @param saddr address of an existing l2addr_154 object
 * @return See the pktype_t type
 */

pktype_t l2net_154::recv (l2addr **saddr, void *data, int *len)
{
    pktype_t r ;

    r = PK_NONE ;			// no packet received
    while (r == PK_NONE)
    {
	// try to find an already received packet (in the list)
	r = extract_received_packet (saddr, data, len) ;

	if (r == PK_NONE)
	{
	    // No packet received: waits for a complete frame in the buffer
	    // and store it (them) in the list
	    if (read_complete_frame () == -1)
		break ;				// error
	}
    }

    D (D_MESSAGE, "Received packet (" << *len << " bytes)") ;

    return r ;
}

// Explore frame list to find (and remove) a received packet
pktype_t l2net_154::extract_received_packet (l2addr **saddr, void *data, int *len)
{
    pktype_t r = PK_NONE ;
    std::list <casan::l2net_154::frame>::iterator f ;

    f = framelist_.begin () ;
    while (r == PK_NONE && f != framelist_.end ())
    {
	if (f->type == casan::l2net_154::RX_SHORT)
	{
	    // get source address and convert it to a l2addr_154 object
	    char txtaddr [6] ;
	    std::snprintf (txtaddr, sizeof txtaddr, "%2x:%2x", 
		BYTE_LOW (f->rx_short_.saddr), BYTE_HIGH (f->rx_short_.saddr)) ;
	    *saddr = new l2addr_154 (txtaddr) ;

	    // transfer data
	    if (*len >= f->rx_short_.len)
		*len = f->rx_short_.len ;
	    std::memcpy (data, f->rx_short_.data, *len) ;

	    // packet addressed to us?
	    if (f->rx_short_.options & casan::l2net_154::RX_SHORT_OPT_BROADCAST)
		r = PK_BCAST ;
	    else r = PK_ME ;

	    // remove frame from list
	    // std::list <casan::l2net_154::frame>::iterator tmp ;
	    // tmp = f ;
	    // tmp++ ;
	    framelist_.erase (f) ;
	    // f = tmp ;
	}
	else f++ ;
    }
    return r ;
}

// Buffer does not contain a complete valid frame. Read a complete frame
int l2net_154::read_complete_frame (void)
{
    int n = -1 ;
    bool found_at_least_one = false ;

    /*
     * We should never reach the end of this buffer
     */

    while (! found_at_least_one &&
	(n = read (fd_, pbuffer_, BUFLEN - (pbuffer_ - buffer_))) > 0)
    {
	pbuffer_ += n ;

	// we may have read more than one frame:
	// extract them all from the buffer to the list
	while (l2net_154::is_frame_complete ())
	{
	    found_at_least_one = true ;
	    l2net_154::extract_frame_to_list () ;
	}
    }

    return n ;				// -1 <=> error
}

/*
 * This function checks to see if the buffer contains a valid API packet.
 * If no valid packet is found, skip junk bytes and re-analyze buffer.
 * Return:
 * - true: a valid frame has been found
 * - false: the buffer still needs some bytes
 */

bool l2net_154::is_frame_complete (void)
{
    byte *start ;
    bool complete = false ;
    bool invalid = true ;

    start = buffer_ ;			// current start of buffer
    while (invalid)
    {
	int buflen ;
	int framelen ;
	int pktlen ;
	int cksum ;
	int i ;

	buflen = pbuffer_ - start ;

	// discard junk bytes before API start delimiter (start-byte)
	i = 0 ;
	while (i < buflen && start [i] != XBEE_START)
	    i++ ;
	// here, i is the relative (to start) position of the start-byte
	// or i >= buflen if not found

	// has a start-byte been found?
	if (i >= buflen)		// no
	{
	    pbuffer_ = buffer_ ;	// reset buffer to empty
	    break ;			// still not complete
	}

	// align start byte at the buffer start
	if (buffer_ != start + i)
	{
	    buflen -= i ;
	    std::memcpy (buffer_, start + i, buflen) ;
	    pbuffer_ = buffer_ + buflen ;
	}

	// do we have enough stuff to extract frame length?
	if (buflen < XBEE_MIN_FRAME_SIZE)
	    break ;			// still not complete

	// does frame length seem realistic ?
	framelen = (buffer_ [1] << 8) | buffer_ [2] ;
	if (framelen > XBEE_MAX_FRAME_SIZE)
	{
	    // skip start-byte
	    start = buffer_ + 1 ;
	    continue ;
	}
	    
	// is checksum already inside the buffer?
	pktlen = framelen + 4 ;
	if (pktlen > buflen)
	    break ;			// still not complete

	// compare checksum
	cksum = compute_checksum (buffer_) ;
	if (cksum != buffer_ [pktlen - 1])
	{
	    // skip start-byte
	    start = buffer_ + 1 ;
	    continue ;
	}

	// if we get there, a valid frame has been found
	invalid = false ;
	complete = true ;
    }

    return complete ;
}

void l2net_154::extract_frame_to_list (void)
{
    int framelen ;
    int pktlen ;
    struct frame f ;
    
    framelen = (buffer_ [1] << 8) | buffer_ [2] ;
    pktlen = framelen + 4 ;

    f.type = (enum frame_type) buffer_ [3] ;
    switch (buffer_ [3])
    {
	case l2net_154::TX_STATUS :
	    f.tx_status_.frame_id = buffer_ [4] ;
	    f.tx_status_.status = buffer_ [5] ;
	    break ;
	case l2net_154::RX_SHORT :
	    f.rx_short_.saddr = (buffer_ [4] << 8) | buffer_ [5] ;
	    f.rx_short_.len = framelen - 5 ;
	    std::memcpy (f.rx_short_.data, buffer_ + 8, f.rx_short_.len) ;
	    f.rx_short_.rssi = buffer_ [6] ;
	    f.rx_short_.options = buffer_ [7] ;
	    break ;
	default :
	    std::cerr << "PKT API " << (int) buffer_ [3] << " unrecognized\n" ;
	    break ;
    }
    framelist_.push_back (f) ;

    // remove extracted packet from buffer
    pbuffer_ -=  pktlen ;
    if (pbuffer_ - buffer_ > 0)		// buffer still contains more bytes
	std::memcpy (buffer_, buffer_ + pktlen, pbuffer_ - buffer_) ;
}

}					// end of namespace casan
